collector/dhcp.go

```diff
@@ -0,0 +1,401 @@
+
+package collector
+
+import (
+	"github.com/prometheus/client_golang/prometheus"
+)
+
+func init() {
+	registerCollector("dhcp", NewDhcpCollector, "DHCP Server")
+}
+
+// A DhcpCollector is a Prometheus collector for both WMI and perflib DHCP metrics
+type DhcpCollector struct {
+	PacketsReceivedTotal                             *prometheus.Desc
+	DuplicatesDroppedTotal                           *prometheus.Desc
+	PacketsExpiredTotal                              *prometheus.Desc
+	MillisecondsPerPacketAvg                         *prometheus.Desc
+	ActiveQueueLength                                *prometheus.Desc
+	ConflictCheckQueueLength                         *prometheus.Desc
+	DiscoversTotal                                   *prometheus.Desc
+	OffersTotal                                      *prometheus.Desc
+	RequestsTotal                                    *prometheus.Desc
+	InformsTotal                                     *prometheus.Desc
+	AcksTotal                                        *prometheus.Desc
+	NacksTotal                                       *prometheus.Desc
+	DeclinesTotal                                    *prometheus.Desc
+	ReleasesTotal                                    *prometheus.Desc
+	OfferQueueLength                                 *prometheus.Desc
+	DeniedDueToMatch                                 *prometheus.Desc
+	DeniedDueToNonmatch                              *prometheus.Desc
+	FailoverBndupdSentTotal                          *prometheus.Desc
+	FailoverBndupdReceivedTotal                      *prometheus.Desc
+	FailoverBndackSentTotal                          *prometheus.Desc
+	FailoverBndackReceivedTotal                      *prometheus.Desc
+	FailoverBndupdPendingOutboundQueue               *prometheus.Desc
+	FailoverTransitionsCommunicationinterruptedState *prometheus.Desc
+	FailoverTransitionsPartnerdownState              *prometheus.Desc
+	FailoverTransitionsRecoverState                  *prometheus.Desc
+	FailoverBndupdDropped                            *prometheus.Desc
+}
+
+func NewDhcpCollector() (Collector, error) {
+	const subsystem = "dhcp"
+
+	return &DhcpCollector{
+		PacketsReceivedTotal: prometheus.NewDesc(
+			prometheus.BuildFQName(Namespace, subsystem, "packets_received_total"),
+			"Total number of packets received by the DHCP server (PacketsReceivedTotal)",
+			nil,
+			nil,
+		),
+		DuplicatesDroppedTotal: prometheus.NewDesc(
+			prometheus.BuildFQName(Namespace, subsystem, "duplicates_dropped_total"),
+			"Total number of duplicate packets received by the DHCP server (DuplicatesDroppedTotal)",
+			nil,
+			nil,
+		),
+		PacketsExpiredTotal: prometheus.NewDesc(
+			prometheus.BuildFQName(Namespace, subsystem, "packets_expired_total"),
+			"Total number of packets expired in the DHCP server message queue (PacketsExpiredTotal)",
+			nil,
+			nil,
+		),
+		MillisecondsPerPacketAvg: prometheus.NewDesc(
+			prometheus.BuildFQName(Namespace, subsystem, "milliseconds_per_packet_avg"),
```

Comment:
Is this an actual average, or a counter pair of total time spent and total packets processed? 
Either way, we should expose seconds instead of milliseconds

Reply:
After running the collector for a week, the metric has **not** changed once with a value of '13' and '16' for the primary and second DHCP servers, respectively.
I've omitted the metric from the collector.